### DISCUSSION TITLE
S3 upload of photos one by one + temporaryUrl(): instable URLs, no browser caching possible

### DISCUSSION THREAD

**Discussion by levu42** [livewire/livewire#5036] [7. Bugs (archived)]
I had a scenario recently that shows a limitation of `temporaryUrl()` in real life usage. I also wrote some custom code to circumvent it but the more I think about it, the more I consider it something that should be fixed inside Livewire.

The scenario:

A website that allows users to upload a bunch of pictures, my use cases go up to "200 photos of 8MB each" => 1.6GB. The connected storage is an S3 storage, so Livewire only supports uploading one picture at a time. I wrote some custom JS that just looped over the files selected in a `<input type="file">` and calls `$wire.upload(...)` for each of them.

In the template I had a section to preview what you've uploaded and because photos got uploaded one by one, the template got re-rendered after each photo upload. That way, you could watch the progress of the uploads while uploading it, since in my use case that could easily be 5-15min. The relevant code is in really simple form:

```blade
@foreach($this->photos as $photo)
  <img src="{{ $photo->temporaryUrl() }}" class="some-thumbnail-class">
@endforeach
```

Now, I discovered terrible scalability, network usage scaling something that looked like O(n²) with number of photos uploaded and after some investigation I found the reason:

* `temporaryUrl()` generates each time a new URL, not allowing the browser to cache the picture
* also of course a problem was that `temporaryUrl()` gave me a link to the full size image and not a thumbnail

What was the effect of the combination of those? Let's say I'm uploading 5 photos of 10MB each:

* after the 1st photo uploaded: browser downloads 1 photo for preview => 10MB
* after the 2nd photo uploaded: browser downloads 2 photos for preview => +20MB => 30MB
* after the 3rd photo uploaded: browser downloads 3 photos for preview => +30MB => 60MB
* after the 4th photo uploaded: browser downloads 4 photos for preview => +40MB => 100MB
* after the 5th photo uploaded: browser downloads 5 photos for preview => +50MB => 150MB

Solution I found:

The solution I found was to make my own preview route & link generation to the preview route. Apart from delivering scaled down thumbnails, I also changed the link generation to have the expiry not  `now()->addMinutes(30)` like [Livewire does currently](https://github.com/livewire/livewire/blob/40519653186241972e42713ad2b38da7abf1f91a/src/TemporaryUploadedFile.php#L92) but instead use `now()->addMinutes(30)->endOfHour()` which means that subsequent requests will generate a URL that's always stable for one hour allowing browsers to cache.

Now, what do I expect changed in Livewire?

First of all, I think change of adding `->endOfHour()` is a change that should be included in Livewire and I would consider it a bug that the URL is currently regenerated on every component update.

As a feature request, I would also like to see some support for having temporary preview URLs being scaled down as thumbnails, either by Livewire having that functionality directly internally or by allowing some way of extending `->temporaryUrl()` on your own, however that fits best into the overall development direction of Livewire.

**Comment by joshhanley** [comment]
@levu42 ah yeah that makes sense, all PR's are welcome! 🙂

**Reply by levu42** [reply]
@joshhanley I'll create one in the next 1-2 weeks for the ->endOfHour() part!

Do you have any thoughts on the thumbnail generation part? Should that be something included in Livewire?

**Reply by joshhanley** [reply]
@levu42 thanks!

Personally, I don't think Livewire should handle the thumbnail generation as every app could be different on how it should be handled (sizes, encoding, etc), so I think providing a way to hook into the temporary URL would be a better option, so people can customise how they like

**Reply by levu42** [reply]
@joshhanley I now re-submitted the PR as #6669 targeted at v3, after you closed the original one I created with the reason being that you try to not accept v2 PRs while working on v3.
